### PR TITLE
Add support for tabs

### LINKED ISSUE #3237: Null object tried to be parsed as JSON
## Exception
* __User Action:__ requested channel
* __Request:__ https://www.youtube.com/channel/UCRedgf5rdWzUCIP5HGwU84w
* __Content Language:__ GB
* __Service:__ YouTube
* __Version:__ 0.18.6
* __OS:__ Linux Android 9 - 28

<details><summary><b>Crash log</b></summary><p>

```
java.lang.NullPointerException: Attempt to invoke virtual method 'com.grack.nanojson.JsonObject com.grack.nanojson.JsonObject.getObject(java.lang.String)' on a null object reference
	at org.schabi.newpipe.extractor.services.youtube.extractors.YoutubeChannelExtractor.getPage(YoutubeChannelExtractor.java:200)
	at org.schabi.newpipe.extractor.channel.ChannelInfo.getMoreItems(ChannelInfo.java:53)
	at org.schabi.newpipe.util.ExtractorHelper.lambda$getMoreChannelItems$5(ExtractorHelper.java:131)
	at org.schabi.newpipe.util.-$$Lambda$ExtractorHelper$W_-kT4Q2PcaBmHMRES2jl1YH31Y.call(Unknown Source:6)
	at io.reactivex.internal.operators.single.SingleFromCallable.subscribeActual(SingleFromCallable.java:44)
	at io.reactivex.Single.subscribe(Single.java:3438)
	at io.reactivex.internal.operators.single.SingleSubscribeOn$SubscribeOnObserver.run(SingleSubscribeOn.java:89)
	at io.reactivex.Scheduler$DisposeTask.run(Scheduler.java:578)
	at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
	at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:764)

```
</p></details>
<hr>
While browsing 'whats new' section, headed to the last video in my channel, then clicked my channel to browse all videos, then scrolled up and down, while browsing down (I guess this time the list gets populated) there was an error, which I've pasted (formatted) above.
thanks.

### PR DESCRIPTION
**Pull request by wb9688**
- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.

Currently it looks like the following:

Videos tab | Playlists tab
--- | ---
Screenshot of Videos tab | Screenshot of Playlists tab

You could download a test APK [here](https://github.com/TeamNewPipe/NewPipe/files/4354794/app-debug.apk.zip).

To do:

- [x] Show the subscribe count and button
- [x] Show placeholder for avatar and banner while loading or when there is none
- [x] Show 'No items' instead of 'No videos'
- [x] Make the queue working again (~though I've no idea how to do that as a tab could contain both streams and playlists~ just for the stream-only tabs)
- [x] Show tab names at the top instead of dots at the bottom
- [x] Clean up the code
- [x] Fix crash when opening item in channel on homepage
- [x] Localize tab names
- [x] Fix location of loading animation and network error

Fixes #3237

Also see TeamNewPipe/NewPipeExtractor#279.

### CODE REVIEW

**Comment by Stypox** [comment]
I am not sure it is a good idea to use those bubbles on the bottom to switch to different tabs. I think a pager like in the MainActivity would suit better, since in the future we could add new tabs for channels (e.g. related/suggested channels, channel description, ...), and having a small label for every tab would be  imo.

**Comment by wb9688** [comment]
@Stypox: I agree. What do you mean with 'a small label' though? Just the name of the tab (e.g. "Videos" or "Popular tracks")?

File: app/src/main/java/org/schabi/newpipe/fragments/list/channel/ChannelTabFragment.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        if (savedState != null) {
+            Serializable serializable = savedState.getSerializable(INFO_KEY);
+            if (serializable instanceof ChannelTabInfo){
+                this.relatedStreamInfo = (ChannelTabInfo) serializable;
+            }
+        }
+    }
+
+    @Override
+    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String s) {}
```
Comment (Stypox) [inline app/src/main/java/org/schabi/newpipe/fragments/list/channel/ChannelTabFragment.java]: Also here, remove if not needed

**Comment by Stypox** [comment]
> @Stypox: I agree. What exactly do you mean with 'a small label' though? Just the name of the tab (e.g. 'Videos' or 'Popular tracks')?

Yeah ;-)

**Comment by heberjeur** [comment]
Great job thanks, can you add the publish date? 
Screenshot_20200309-103505

**Comment by B0pol** [comment]
No you can't.
See by yourself on YouTube website, there are no upload dates.

**Comment by heberjeur** [comment]
I found this error
https://youtu.be/90wuJhtQlqE

File: app/src/main/res/layout/fragment_channel_detail.xml [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+					android:layout_marginTop="@dimen/video_item_detail_error_panel_margin"
+					android:visibility="gone"
+					tools:visibility="visible" />
+
+			</RelativeLayout>
+
+			<com.google.android.material.tabs.TabLayout
+				android:id="@+id/tablayout"
+				android:layout_width="match_parent"
+				android:layout_height="wrap_content" />
```
Comment (Stypox) [inline app/src/main/res/layout/fragment_channel_detail.xml]:
```suggestion
			<com.google.android.material.tabs.TabLayout
				android:id="@+id/tablayout"
				android:layout_width="match_parent"
				android:layout_height="wrap_content">
				
				<com.google.android.material.tabs.TabItem
                    android:text="@string/tab_text" />
				
			</com.google.android.material.tabs.TabLayout>
```

**Comment by wb9688** [comment]
I updated the screenshots and test APK.

**Comment by opusforlife2** [comment]
Would you please rename the package for adjacent installation with other apks?

**Comment by B0pol** [comment]
> Would you please rename the package for adjacent installation with other apks?

Rebase because it will also include fixed feed

File: app/src/main/res/values/strings.xml [hunk 3]
```diff
@@ -599,4 +601,12 @@
     <plurals name="dynamic_seek_duration_description">
         <item quantity="other">%s seconds</item>
     </plurals>
+    <!-- Channel tabs -->
+    <string name="channel_tab_videos">Videos</string>
```
Comment (B0pol) [inline app/src/main/res/values/strings.xml]:
```suggestion
```
use ``R.string.videos_string`` instead

File: app/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -599,4 +601,12 @@
...
+    <string name="channel_tab_playlists">Playlists</string>
+    <string name="channel_tab_popular_tracks">Popular tracks</string>
+    <string name="channel_tab_tracks">Tracks</string>
+    <string name="channel_tab_albums">Albums</string>
+    <string name="channel_tab_reposts">Reposts</string>
+    <string name="channel_tab_events">Events</string>
```
Comment (B0pol) [inline app/src/main/res/values/strings.xml]:
```suggestion
```
use ``R.string.events`` instead

File: app/src/main/res/values/strings.xml [hunk 5]
```diff
@@ -599,4 +601,12 @@
     <plurals name="dynamic_seek_duration_description">
         <item quantity="other">%s seconds</item>
     </plurals>
+    <!-- Channel tabs -->
+    <string name="channel_tab_videos">Videos</string>
+    <string name="channel_tab_playlists">Playlists</string>
+    <string name="channel_tab_popular_tracks">Popular tracks</string>
+    <string name="channel_tab_tracks">Tracks</string>
```
Comment (B0pol) [inline app/src/main/res/values/strings.xml]:
```suggestion
```
use R.string.tracks instead

File: app/src/main/res/values/strings.xml [hunk 6]
```diff
@@ -599,4 +601,12 @@
     <plurals name="dynamic_seek_duration_description">
         <item quantity="other">%s seconds</item>
     </plurals>
+    <!-- Channel tabs -->
+    <string name="channel_tab_videos">Videos</string>
+    <string name="channel_tab_playlists">Playlists</string>
```
Comment (B0pol) [inline app/src/main/res/values/strings.xml]:
```suggestion
```
use ``R.string.playlists`` instead

File: app/src/main/java/org/schabi/newpipe/local/subscription/SubscriptionManager.kt [hunk 7]
```diff
@@ -27,7 +27,7 @@ class SubscriptionManager(context: Context) {
 
         database.runInTransaction {
             infoList.forEachIndexed { index, info ->
-                feedDatabaseManager.upsertAll(listEntities[index].uid, info.relatedItems)
+                feedDatabaseManager.upsertAll(listEntities[index].uid, info.tabs[0].relatedItems as List<StreamInfoItem>)
```
Comment (wb9688) [inline app/src/main/java/org/schabi/newpipe/local/subscription/SubscriptionManager.kt]: Is there a better solution for this?

File: app/src/main/java/org/schabi/newpipe/fragments/detail/ChannelDetailFragment.java [hunk 8]
```diff
@@ -0,0 +1,557 @@
...
+    //////////////////////////////////////////////////////////////////////////*/
+
+    @Override
+    protected boolean onError(Throwable exception) {
+        if (super.onError(exception)) return true;
+
+        else if (exception instanceof ContentNotAvailableException) {
+            showError(getString(R.string.content_not_available), false);
+        } else {
+            int errorId = exception instanceof YoutubeStreamExtractor.DecryptException
```
[the remark this answers is not in the document]
Comment (wb9688) [reply app/src/main/java/org/schabi/newpipe/fragments/detail/ChannelDetailFragment.java]: This whole thing is stupid as it's meant for streams, not channels lol

**Comment by wb9688** [comment]
I just rebased and updated the APK.

File: app/src/main/java/org/schabi/newpipe/player/playqueue/ChannelPlayQueue.java [hunk 9]
```diff
@@ -39,9 +32,9 @@ public void fetch() {
             ExtractorHelper.getChannelInfo(this.serviceId, this.baseUrl, false)
                     .subscribeOn(Schedulers.io())
                     .observeOn(AndroidSchedulers.mainThread())
-                    .subscribe(getHeadListObserver());
+                    .subscribe((Consumer<? super ChannelInfo>) getHeadListObserver());
```
Comment (wb9688) [inline app/src/main/java/org/schabi/newpipe/player/playqueue/ChannelPlayQueue.java]: I don't think I did this part correctly and I'm not sure what this even is supposed to do, though I don't think it breaks anything when I remove it.

**Comment by wb9688** [comment]
@TobiGr: I think this PR is almost finished now. I highlighted 2 issues above, plus I need to make the playlist controls not sticky. @B0pol also discovered some SoundCloud issue, where it sometimes stops working completely for a few minutes, which I've reproduced once, but I'm not sure what causes it.

**Comment by wb9688** [comment]
Another bug found by @B0pol: when a channel only has one tab (i.e. media.ccc.de or PeerTube) and you press back, it crashes.

**Comment by opusforlife2** [comment]
Something is wrong. The feed update process is broken. It doesn't load even a single one successfully. Tapping on Report doesn't even show the errors. It just closes the app. I'm not sure if that's caused by this PR or mauricio's code.

**Comment by wb9688** [comment]
@opusforlife2: I could confirm that, however it shows the errors for me. I'll look into that.

**Comment by wb9688** [comment]
@opusforlife2: I fixed your issue and rebased it (again). I'll upload the new APK in a minute.

**Comment by opusforlife2** [comment]
Confirmed fixed.

**Comment by opusforlife2** [comment]
Another bug, but it is also present in cool-student's PR. When you open the app through a link from another app, the name of the app "Newpipe tabs" flashes for a moment at the place where the resolution drop down is. For his PR, it's "Newpipe notification", naturally. This doesn't happen in the release apk.

Which means it's something in the dev branch causing the issue.

**Comment by wb9688** [comment]
@opusforlife2: Please just create a new issue for that as that isn't caused by this PR.

**Comment by opusforlife2** [comment]
That... is a very good point. Not sure what I was thinking. Brain fart.

File: app/src/main/java/org/schabi/newpipe/local/feed/service/FeedLoadService.kt [hunk 10]
```diff
@@ -211,7 +211,7 @@ class FeedLoadService : Service() {
                         } else {
                             ExtractorHelper
                                     .getChannelInfo(subscriptionEntity.serviceId, subscriptionEntity.url, true)
-                                    .blockingGet()
+                                    .blockingGet().tabs[0]
```
Comment (mauriciocolli) [inline app/src/main/java/org/schabi/newpipe/local/feed/service/FeedLoadService.kt]: See https://github.com/TeamNewPipe/NewPipeExtractor/pull/279#discussion_r394446468.

**Comment by wb9688** [comment]
I rebased and implemented lazy-loading. I've updated the screenshots and APK in the OP.

@opusforlife2: Feel free to test.

File: app/src/main/java/org/schabi/newpipe/fragments/list/channel/ChannelTabFragment.java [hunk 11]
```diff
@@ -126,6 +127,10 @@ public void handleResult(@NonNull ChannelTabInfo result) {
                 headerBackgroundButton.setOnClickListener(
                         view -> NavigationHelper.playOnBackgroundPlayer(activity, getPlayQueue(0), false));
         }
+
+        if (channelTabInfo.isUsedForFeed()) {
+            new SubscriptionManager(activity).updateChannelTabInfo(currentInfo);
```
Comment (wb9688) [inline app/src/main/java/org/schabi/newpipe/fragments/list/channel/ChannelTabFragment.java]: Should I use RX here somehow?

**Comment by opusforlife2** [comment]
Wow! I came here to report a bug in the previous apk, but you seem to have fixed it in the latest one, since I can no longer reproduce it.

The bug was that the 'Subscribed' button would flicker between toggling off and on sometimes, when scrolling up fast or switching tabs. I can't pinpoint which commit fixed it, though.

**Comment by opusforlife2** [comment]
Are Videos and Playlists the only two tabs implemented for now or am I just unable to find a channel with other tabs?

**Comment by wb9688** [comment]
@opusforlife2: Those are indeed the only 2 tabs that are currently implemented for YouTube.

**Comment by B0pol** [comment]
> Are Videos and Playlists the only two tabs implemented for now or am I just unable to find a channel with other tabs?

on YouTube, yes. Go on SoundCloud

**Comment by opusforlife2** [comment]
Error when switching tabs on a SoundCloud user page. 

Also, why does Newpipe lose the backstack when you open the error report page? Isn't that a bug?

## Exception
* __User Action:__ requested channel
* __Request:__ https://soundcloud.com/eminemofficial
* __Content Language:__ en_US
* __Service:__ SoundCloud
* __Version:__ 0.18.6
* __OS:__ Linux Android 9 - 28


<details><summary><b>Crash log</b></summary><p>

```
org.schabi.newpipe.extractor.exceptions.ExtractionException: Could not get next page
	at org.schabi.newpipe.extractor.services.soundcloud.SoundcloudChannelAlbumsExtractor.computeNextPageAndGetPlaylists(SoundcloudChannelAlbumsExtractor.java:61)
	at org.schabi.newpipe.extractor.services.soundcloud.SoundcloudChannelAlbumsExtractor.getInitialPage(SoundcloudChannelAlbumsExtractor.java:37)
	at org.schabi.newpipe.extractor.utils.ExtractorHelper.getItemsPageOrLogError(ExtractorHelper.java:19)
	at org.schabi.newpipe.extractor.channel.ChannelTabInfo.loadTab(ChannelTabInfo.java:67)
	at org.schabi.newpipe.fragments.list.channel.ChannelTabFragment.lambda$loadResult$0$ChannelTabFragment(ChannelTabFragment.java:84)
	at org.schabi.newpipe.fragments.list.channel.-$$Lambda$ChannelTabFragment$WuJ37aw-7UKTbW8f72jXH9kxJGI.call(Unknown Source:2)
	at io.reactivex.internal.operators.single.SingleFromCallable.subscribeActual(SingleFromCallable.java:44)
	at io.reactivex.Single.subscribe(Single.java:3438)
	at io.reactivex.internal.operators.single.SingleSubscribeOn$SubscribeOnObserver.run(SingleSubscribeOn.java:89)
	at io.reactivex.Scheduler$DisposeTask.run(Scheduler.java:578)
	at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
	at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:764)
Caused by: java.io.InterruptedIOException: interrupted
	at okio.Timeout.throwIfReached(Timeout.java:146)
	at okio.Okio$1.write(Okio.java:76)
	at okio.AsyncTimeout$1.write(AsyncTimeout.java:180)
	at okio.RealBufferedSink.flush(RealBufferedSink.java:224)
	at okhttp3.internal.http1.Http1Codec.finishRequest(Http1Codec.java:166)
	at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:84)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at com.facebook.stetho.okhttp3.StethoInterceptor.intercept(StethoInterceptor.java:56)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:127)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:257)
	at okhttp3.RealCall.execute(RealCall.java:93)
	at org.schabi.newpipe.DownloaderImpl.execute(DownloaderImpl.java:159)
	at org.schabi.newpipe.extractor.downloader.Downloader.get(Downloader.java:70)
	at org.schabi.newpipe.extractor.downloader.Downloader.get(Downloader.java:42)
	at org.schabi.newpipe.extractor.services.soundcloud.SoundcloudParsingHelper.getPlaylistsFromApi(SoundcloudParsingHelper.java:312)
	at org.schabi.newpipe.extractor.services.soundcloud.SoundcloudParsingHelper.getPlaylistsFromApiMinItems(SoundcloudParsingHelper.java:300)
	at org.schabi.newpipe.extractor.services.soundcloud.SoundcloudChannelAlbumsExtractor.computeNextPageAndGetPlaylists(SoundcloudChannelAlbumsExtractor.java:59)
	... 16 more

```
</p></details>
<hr>

**Comment by wb9688** [comment]
@opusforlife2: I can't reproduce that error.

About the backstack: I've no idea. I had code related to that inherited from the VideoDetailFragment, but that seemed to cause crashes here, so I removed that code, which seemed to work.

**Comment by opusforlife2** [comment]
Keep opening some User pages. You'll eventually hit it. I originally got it with an unofficial Daft Punk page, but then it stopped happening. I thought it was a one off, but then it happened here.

What basically happens is that Newpipe is either able to load the tabs properly, in which case the error is unlikely to appear at all. But if it takes time to load the first tab, chances are that one of the other ones will give the error.

For example, when I hit the error with the Eminem page, the Tracks tab was empty. But when I opened it again a few times, Tracks was populated.

**Comment by Stypox** [comment]
@opusforlife2 for the backstack: that's not something that can be prevented, I think. When the app crashes... it crashes. The Android is instructed to call a function inside of the app to handle the information about the crash, but everything has already been destroyed.

**Comment by opusforlife2** [comment]
@Stypox I'm not talking about the case where the app crashes and then automatically shows the error page. I'm talking about when you're happily browsing, then you see an error at the bottom, which you can choose not to open and continue browsing without any issue. But if you open it, you lose the backstack. I don't see why that should be the case. Maybe there's a way to code the page so that if it is opened manually by the user, the backstack shouldn't be discarded?

Also, the button being called "Report" is a bit misleading. It should be called "View" instead. You get the option to report the error only _after_ viewing it.

**Comment by njmdietrich** [comment]
When opening a playlist on YouTube from the playlists tab and then pressing the back button to return to the tab, it opens on the video tab. While the scrolling position of the playlists tab is preserved so switching back to it doesn't cause any issues, the expected behaviour would be to return to the playlist tab immediately.

**Comment by opusforlife2** [comment]
@nictanghe Follow this for playlists in channels.

**Comment by NicTanghe** [comment]
Does this anly work for playlists with channels own videos or also playlists with others videos?
Also thx for pointing me here. Altyd grappig om mensen Nederland s te zien gebruiken.

**Comment by opusforlife2** [comment]
@NicTanghe Probably the channel's own playlists. Can't remember for sure.

**Comment by BruceMustache** [comment]
What happens with this PR?

**Comment by LostWanita21** [comment]
Curious about the PR as well, checked if there were issues filed for it and found this, not sure if any things come from it.

**Comment by MD77MD** [comment]
any hope for this to be merged

**Comment by alittlebitofit** [comment]
Is the apk still updated?
Because I see the loading indicator running forever.
And it says "No Results" when I try to search.

**Comment by heberjeur** [comment]
@TiA4f8R why?, isn't necessary to add support for tabs ?

**Comment by Stypox** [comment]
Yes it is, since there is an issue open about it which has received endorsements from devs. But nobody is working on this at the moment and the original creator of this PR dropped his work. If somebody wants to work on this, contributions are well accepted.

**Pull request closed without merging**